narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,201 @@
+        })
+    }
+
+    async fn start_dkg(&self, tx_system_messages: &Sender<SystemMessage>) {
+        let msg = self.party.create_message(&mut rand::thread_rng());
+        info!("random beacon: sending DKG Message: {msg:?}");
+        let _ = tx_system_messages
+            .send(SystemMessage::DkgMessage(msg))
+            .await;
+    }
+
+    fn add_message(&mut self, msg: dkg::Message<PkG, EncG>) {
+        if !self.shares.is_empty() {
+            // We've already sent a `Confirmation`, so we can't add any more messages.
+            return;
+        }
+        self.messages.push(msg.clone());
+        match self.party.process_message(msg, &mut rand::thread_rng()) {
+            Ok(processed) => {
+                self.processed_messages.push(processed);
+            }
+            Err(err) => {
+                debug!("error while processing randomness DKG message: {err:?}");
+            }
+        }
+    }
+
+    fn add_confirmation(&mut self, conf: dkg::Confirmation<EncG>) {
+        self.confirmations.push(conf)
+    }
+
+    // Generates the next SystemMessage needed to advance the random beacon protocol, if possible,
+    // and sends it to the proposer.
+    async fn advance(&mut self, tx_system_messages: &Sender<SystemMessage>) {
+        // Once we have enough ProcessedMessages, send a Confirmation.
+        if self.shares.is_empty() && !self.processed_messages.is_empty() {
+            match self.party.merge(&self.processed_messages) {
+                Ok((shares, conf)) => {
+                    info!(
+                        "random beacon: sending DKG Confirmation with {} complaints",
+                        conf.complaints.len()
+                    );
+                    self.shares = shares;
+                    let _ = tx_system_messages
+                        .send(SystemMessage::DkgConfirmation(conf))
+                        .await;
+                }
+                Err(fastcrypto::error::FastCryptoError::InputTooShort(_)) => (), // wait for more input
+                Err(e) => error!("Error while merging randomness DKG messages: {e:?}"),
+            }
+        }
+
+        // Once we have enough Confirmations, process them and update shares.
+        if self.dkg_output.is_none() && !self.confirmations.is_empty() {
+            match self.party.process_confirmations(
+                &self.messages,
+                &self.confirmations,
+                self.shares.clone(),
+                self.party.t() * 2 - 1, // t==f+1, we want 2f+1
+                &mut rand::thread_rng(),
+            ) {
+                Ok(shares) => {
+                    self.dkg_output = Some(self.party.aggregate(&self.messages, shares));
+                    info!(
+                        "random beacon: DKG complete with Output {:?}",
+                        self.dkg_output
+                    );
+                }
+                Err(fastcrypto::error::FastCryptoError::InputTooShort(_)) => (), // wait for more input
```

Comment:
also here, let's add a TODO to use a more specific error

Reply:
done